### PR TITLE
perf(replays): query directly on replay_id uuid instead of stripped s…

### PR DESCRIPTION
**Pull request by JoshFerge**
Using our base query, it's shown that removing the string conversion / dash stripping on the replay_id column improves memory usage by ~26% and speed by ~21%. Queries pasted in at bottom.

This PR creates a new `Field` `UUIDField` as we have to do validation / conversion prior to sending our query to clickhouse, and our minimum clickhouse version doesn't have useful helpful functions so we have to work around.

We also now have to strip dashes on replay_ids in the post processing of our queries. We could do the same for error_ids and trace_ids in a future PRs if we want that optimization on those fields as well. 

```
SET send_logs_level = 'trace'
```

```
SELECT　    project_id AS _snuba_project_id,　    replaceAll (toString (replay_id), '-', '') AS _snuba_replay_id,　    max(timestamp AS _snuba_timestamp) AS _snuba_finished_at,　    sum(length(error_ids AS _snuba_error_ids)) AS _snuba_count_errors,　    groupArray (1) (environment AS _snuba_environment)[1] AS _snuba_agg_environment,　    _snuba_replay_id,　    notEmpty (groupArray (is_archived AS _snuba_is_archived)) AS _snuba_isArchived　FROM　    replays_local　WHERE (_snuba_project_id IN[11276])　    AND (_snuba_timestamp < toDateTime ('2023-03-11T00:30:34', 'Universal'))　    AND (_snuba_timestamp >= toDateTime ('2022-03-10T00:30:34', 'Universal'))　GROUP BY　    _snuba_project_id,　    _snuba_replay_id　HAVING (min(segment_id AS _snuba_segment_id) = 0)　AND (_snuba_finished_at < toDateTime ('2023-03-11T00:30:34', 'Universal'))　AND (_snuba_isArchived = 0)　ORDER BY　    _snuba_count_errors ASC　LIMIT 0,　10

Peak memory usage (for query): 612.65 MiB.
10 rows in set. Elapsed: 2.912 sec. Processed 7.94 million rows, 341.42 MB (2.73 million rows/s., 117.25 MB/s.)
```

```
SELECT　    project_id AS _snuba_project_id,　    replay_id,　    max(timestamp AS _snuba_timestamp) AS _snuba_finished_at,　    sum(length(error_ids AS _snuba_error_ids)) AS _snuba_count_errors,　    groupArray (1) (environment AS _snuba_environment)[1] AS _snuba_agg_environment,　    notEmpty (groupArray (is_archived AS _snuba_is_archived)) AS _snuba_isArchived　FROM　    replays_local　WHERE (_snuba_project_id IN[11276])　    AND (_snuba_timestamp < toDateTime ('2023-03-11T00:30:34', 'Universal'))　    AND (_snuba_timestamp >= toDateTime ('2022-03-10T00:30:34', 'Universal'))　GROUP BY　    _snuba_project_id,　    replay_id　HAVING (min(segment_id AS _snuba_segment_id) = 0)　AND (_snuba_finished_at < toDateTime ('2023-03-11T00:30:34', 'Universal'))　AND (_snuba_isArchived = 0)　ORDER BY　    _snuba_count_errors ASC　LIMIT 0,　10

MemoryTracker: Peak memory usage (for query): 448.65 MiB
10 rows in set. Elapsed: 1.711 sec. Processed 7.94 million rows, 341.42 MB (4.64 million rows/s., 199.60 MB/s.)
```

### CODE REVIEW

**Comment by JoshFerge** [comment]
needs https://github.com/getsentry/getsentry/pull/9792 to get tests to pass

File: src/sentry/replays/lib/query.py [hunk 1]
```diff
@@ -92,6 +92,33 @@ def as_condition(
...
+            uuids = _transform_uuids(value)
+            if uuids is None:
+                return Condition(Function("identity", parameters=[1]), Op.EQ, 2)
+            value = [Function("toUUID", parameters=[v]) for v in value]
+        else:
+            uuid = _transform_uuids([value])
+            if uuid is None:
+                return Condition(Function("identity", parameters=[1]), Op.EQ, 2)
+
+            value = Function("toUUID", parameters=uuid)
```
Comment (cmanallen) [inline src/sentry/replays/lib/query.py:117]:
```suggestion
            value = Function("toUUID", parameters=uuid[0])
```

Needs to use the value in the first index.  Not the best API I know.

**Review by cmanallen — COMMENTED** [review]
Great job on this!  This will be a great improvement.

File: src/sentry/replays/lib/query.py (hunk 1 above)
Comment (JoshFerge) [reply src/sentry/replays/lib/query.py:117]: I think the current code is correct: value will only be one element, so `transform_uuids` will return a list w/ one element. `parameters` expects a list, not a scalar value. your suggestion results in:


```
sentry.utils.snuba.QueryExecutionError: DB::Exception: Incorrect number of arguments for function toUUID provided 36 (String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String, String), expected 1 ('Value' : ): While processing SELECT project_id AS _snuba_project_id, replay_id AS _snuba_replay_id, notEmpty(groupArray(is_archived AS _snuba_is_archived)) AS _snuba_isArchived, groupArray(1)(environment AS _snuba_environment)[1] AS _snuba_agg_environment, min(replay_start_timestamp AS _snuba_replay_start_timestamp) AS _snuba_started_at, _snuba_replay_id, max(timestamp AS _snuba_timestamp) AS _snuba_finished_at FROM replays_local WHERE ((_snuba_project_id IN [4551592305098753]) AND (_snuba_timestamp < toDateTime('2023-03-15T19:29:07', 'Universal')) AND (_snuba_timestamp >= toDateTime('2022-12-15T19:29:07', 'Universal'))) AND ((_snuba_replay_id = toUUID('f', '3', '9', '0', '6', '8', 'f', '6', '-', '0', 'f', 'b', 'e', '-', '4', '5', 'a', '0', '-', 'a', '8', '9', '9', '-', 'a', '5', '3', '6', 'e', '3', 'a', '2', '2', 'a', '6', 'f')) OR (_snuba_replay_id = toUUID('2', '8', '1', '7', '3', 'b', '1', '9', '-', '4', '1', '1', '8', '-', '4', '9', '6', '1', '-', '9', '5', '8', '0', '-', '9', 'a', '6', '5', '8', '3', '8', '4', 'a', '8', '1', '7')) OR (_snuba_replay_id = toUUID('e', '8', 'd', 'a', 'e', '8', '4', 'a', '-', '9', 'c', 'a', '9', '-', '4', '6', '0', '8', '-', '8', '1', '6', 'c', '-', '7', '5', '7', '7', '9', 'd', '2', '4', '9', 'a', 'd', '6'))) GROUP BY _snuba_project_id, _snuba_replay_id HAVING ((min(segment_id AS _snuba_segment_id) = 0) AND (_snuba_finished_at < toDateTime('2023-03-15T19:29:07', 'Universal'))) AND (_snuba_isArchived = 0) ORDER BY _snuba_started_at DESC LIMIT 0, 10. Stack trace:
````

**Comment by JoshFerge** [comment]
this is currently blocked as on clickhouse 20.8, our minimum version supported, it seems we cannot supply UUIDs to an in operator. we'd like to use the primary key as we have an index on it, so i'd like to avoid a materialized column. 

I'm wondering if we only make this type of search available via an environment variable, or perhaps just don't allow this type of query?

**Comment by JoshFerge** [comment]
was able to fix after playing around on a clickhouse 20.3 instance for a second -- I didn't need to use the toUUID function, just had to simply ensure the uuid strings had dashes in them.

**Review by cmanallen — APPROVED** [review]

**Pull request merged**